### ISSUE TITLE
Verify that the rest of Shaka's code/functionality matches what I have here.

### ISSUE THREAD

**Issue by Pulseczar1** [Pulseczar1/Shaka-CustomTF-2009#3]
Check that we aren't missing any changes that Shaka, or anyone else, made to the progs code in his 2009 version of CuTF. See what changes Shaka did to the server engine, and make sure those changes are in here, as well.

**Comment by Pulseczar1** [comment]
I worked on this tonight. I haven't made any code changes yet. Using old CustomTF Improvement Project repositories on SourceForge, I determined some things.
The main page for all of the repositories hosted there:
https://sourceforge.net/p/customtf/code/HEAD/tree/
This is Shaka's 2009 CuTF **progs** code, as it was hosted on the Phoenixlabs.org server:
https://sourceforge.net/p/customtf/code/HEAD/tree/prozactf/
The only change I noticed between that code and Shaka's repository on GitHub (the code used here) is that Shaka's bunny hop restriction code was disabled (I believe by Phrostbyte) in the SourceForge repository. I'm going to keep that BH restriction code in tact. I might create a switch to turn it on and off during runtime at some point, though. The BH restriction code seems to work properly on the server, when I test it, using my code, here.

This is Shaka's 2009 CuTF **server engine** code, as it was hosted on the Phoenixlabs.org server:
https://sourceforge.net/p/customtf/code/HEAD/tree/cpqw
In this repository, I compared revision 199, that has Shaka's changes, to the previous revision (184), so that I could see what Shaka did. I found that he put comments in a few files, but never made changes. So, I ignored those changes (since the comments seemed of no value to us, in MVDSV, especially). He figured out a way to read the `.job` field, but ended up never using it. So, I ignored changes dealing with grabbing that value. I found one comment where he talked about not only making changes for spy (see below), but also for thief. My guess is that he wanted to also make it so that invisible thieves are visible to teammates, but he never finished that part. In pmove.c, changes are missing that I believe were added previously to fix random bunny hop failures. I'm not concerned with that for two reasons: his BH restriction code in the progs works just fine, regardless, and MVDSV has long had bunny hopping that works well. So, I'm ignoring the changes in pmove.c.

There are some things in Shaka's engine code that are missing in Often's MVDSV fork, that he calls "PQWSV", which we are using:
1. Spies are supposed to have their `bottomcolor` set to their true color, when viewed by teammates. Teammates only see your `topcolor` changed to the enemy team, letting them know you are a disguised teammate. The code that accomplishes that is in sv_user.c.
2. In sv_phys.c, Shaka doubles `sv_maxvelocity` to 4000, and sets `sv_maxspeed` to 500 from 320. He gives notes about why he does this and they make sense to me to keep. This code I imagine only sets the default values. These things should be set-able from the console or a configuration file. But, since the server engine in this repository is included in with the repository, I think it's fine to make it work specifically for Shaka's 2009 CuTF. So, I'm probably just going to make sure those two variables are high enough in PQWSV, in the code.

To summarize, files still being considered for requiring ported code into this repository are:
1. pmove.c  (probably won't include these changes)
2. sv_phys.c  (probably will include these)
3. sv_user.c  (will need to include this functionality, however I decide to do it, so that friendlies can see when you are disguised)

One final point: I need to make sure that the spy visibility code changes work correctly in Neo mode. In the SourceForge repository, you can see where I added a revision to the engine code (r612 - https://sourceforge.net/p/customtf/code/612/) that fixes a problem with the spy visibility code in Neo mode.

**Comment by Pulseczar1** [comment]
I worked on figuring out where I need to put code changes to make it so that when a player disguises their color, their enemies see their fully-disguised color, but their teammates see their true bottomcolor. Shaka puts the changes for this in `SV_SetInfo_f()`, in CPQWSV. In MVDSV (PQWSV), that function is now called `Cmd_SetInfo_f()`. In that function, Often has put code to disable the function if it's called for the player's team, color, or skin, because we aren't letting clients directly change that information in TF. So, the code that does the needed functionality I think needs to be put in the built-in `setinfo()` function, called `PF_setinfo()` in the engine. At the end of that function, it sends out the information to all clients to report the userinfo change to each client connected to the server. The way Shaka changed the code, before, was to detect when he needed to send different information to certain clients, and do just that. I think I need to put that code at the bottom of `PF_setinfo()`.

I stopped working soon after pasting Shaka's code into `PF_setinfo()`. I made it clear what I pasted in. I was in the process of porting that pasted code to work in MVDSV/PQWSV.

**Comment by Pulseczar1** [comment]
I finished porting all the code. I compiled and ran the server successfully... until I switched to a TF map with teams. start.bsp works fine. When I go to a TF map, likely because there is more than 1 team, the server crashes with `ERROR: SZ_GetSpace: 1/0 is > full buffer size`. It likely has to do with my code, because the new code in `PF_setinfo()` only executes when there is more than one team. I'm pretty sure the problem is in the send network message calls. Should look at call stack in GDB at the time of the crash. I suspect the problem source is actually in the copying of the userinfo data structure that I do, to make a fake userinfo. The problem with doing that is that that structure contains pointers. I believe it contains at least one linked list. If you copy a linked list like that, your copy is going to have pointers pointing to the data of the structure that was copied. I have to build my fake userinfo structure in a different way. I'll see how `_userinfo_ctx_` is built. I'll also look to see if there is a function already written for copying them.

**Comment by Pulseczar1** [comment]
I did away with using a fake userinfo copy, just for storing the bottomcolor. It saves me from having to continue trying to figure out MVDSV's userinfo_ctx stuff. We don't need all that. We just need to send a fake/different bottomcolor to teammates. I've got it working much better now, currently not crashing, and mostly working. It needs a lot of clean-up, though. And I would love to make it so that it always sends the "correct" color, depending on whether it's sending to a teammate or not, whether we are sending out a player's color because they just disguised, or because another player changed team or just joined the server.

In order to send the proper updates to the bottomcolors when a client joins the server, or changes team, and so on, it looks like it would involve the function `SV_FullClientUpdate()` and `svc_updateuserinfo` (which sends all userinfo for a client at once, rather than one key-and-value at a time) and `_userinfoshort_ctx_`.

In its current state, missing the functionality of the previous paragraph, it seems from what I recall of all the changes that Shaka made, that his changes also only updated the bottomcolor for teammates when the teammate disguised their bottomcolor. It did not update when a player joins/leaves the team of the disguised spy. So, I might decide to leave it as is for now, though probably clean the code up some more first. I'll probably create a new issue to address the remaining problem described here, link that issue to this issue somehow, and get to fixing that problem whenever (or someone else can fix it).

**Comment by Pulseczar1** [comment]
I rewrote Shaka's code in the engine, in `PF_setinfo()`. It should function the same, but I prefer how I have it structured. It still has the deficiency mentioned in the previous comment, above. I will create a new Issue for that. Everything seems to be working. So, I'm going to merge this branch into the main branch, and close this Issue, once I do that.

Manually merged this issue/branch using instructions here, so that I could squash all commits in this branch into one commit message that I specify:  https://stackoverflow.com/a/38654139
```
$ git checkout mainBranch
$ git merge featureBranch --squash --no-commit
$ git difftool --cached --dir-diff
$ git commit -m "message"
$ git push origin mainBranch
```
